src/check_presence.c

```diff
@@ -0,0 +1,533 @@
+   * we back off for enough time to allow remote hosts to check in and
+   * remind us they're still there.
+   *
+   * Any host that hasn't checked in by this point can be safely assumed
+   * to have vanished, and will accurately trigger a notification.
+   */
+  struct timespec ts_wait = {0};
+
+  ts_wait.tv_sec = tv_begin.tv_sec + startup_delay + thread_interval;
+  ts_wait.tv_nsec = 0;
+
+  pthread_cond_timedwait(&presence_cond, &presence_lock, &ts_wait);
+
+  while (presence_thread_loop > 0) {
+
+    struct timeval tv_now = {0};
+
+    if (gettimeofday(&tv_now, NULL) < 0) {
+      ERROR(PLUGIN_NAME " plugin: gettimeofday failed: %s", STRERRNO);
+      presence_thread_loop = 0;
+      break;
+    }
+
+    pthread_mutex_unlock(&presence_lock);
+
+    /*
+     * Cleanup thread - have any hosts been gone too long?
+     *
+     * We run with a two second delay between each invocation, and walk
+     * the hosts list looking for hosts that we haven't seen for more
+     * than our threshold.
+     *
+     * Hosts that have gone missing are removed from our list, and a
+     * notification is sent.
+     *
+     * When the host returns, the host will be detected by host_write,
+     * and a notification that the host has been seen will be sent.
+     */
+
+    MDB_txn *txn = NULL;
+
+    int rc = mdb_txn_begin(env, NULL, 0, &txn);
+    if (rc) {
+      ERROR(PLUGIN_NAME " plugin: mdb_txn_begin returned: %s (%d)",
+            mdb_strerror(rc), rc);
+      goto skip;
+    }
+
+    MDB_dbi dbi = {0};
+
+    rc = mdb_dbi_open(txn, NULL, MDB_CREATE, &dbi);
+    if (rc) {
+      ERROR(PLUGIN_NAME " plugin: mdb_dbi_open returned: %s (%d)",
+            mdb_strerror(rc), rc);
+      mdb_txn_abort(txn);
+      goto skip;
+    }
+
+    MDB_cursor *cursor = NULL;
+
+    rc = mdb_cursor_open(txn, dbi, &cursor);
+    if (rc) {
+      ERROR(PLUGIN_NAME " plugin: mdb_cursor_open returned: %s (%d)",
+            mdb_strerror(rc), rc);
+      mdb_txn_abort(txn);
+      goto skip;
+    }
+
+    MDB_val key = {0}, data = {0};
+
+    while ((rc = mdb_cursor_get(cursor, &key, &data, MDB_NEXT)) == 0) {
+
+      /* sanity check - remove any entries that have mismatched sizes,
+       * this might happen after an upgrade.
+       */
+
+      notification_t n = {0};
+
+      if (key.mv_size != sizeof(n.host) ||
+          data.mv_size != sizeof(struct timespec)) {
+
+        rc = mdb_cursor_del(cursor, 0);
+        if (rc) {
+          ERROR(PLUGIN_NAME " plugin: mdb_cursor_del returned: %s (%d)",
+                mdb_strerror(rc), rc);
+          mdb_cursor_close(cursor);
+          mdb_txn_abort(txn);
+          goto skip;
+        }
+
+        continue;
+      }
+
+      /* too old? if so, we've lost the host.
+       *
+       * also - stay silent for the first delay_interval after startup so we
+       * don't send a flurry of stale alerts.
+       *
+       */
+
+      const char *host = key.mv_data;
+
+      struct timeval *tv_then = {0};
```

Comment:
Why do not merge these two lines into one?

Reply:
Assuming you're asking for them to be merged - done.